[WIP] Adds PennyLane QAOA Module Tutorial

Lucaman99 opened the pull request:
**WARNING:** This tutorial will not run until PRs: https://github.com/PennyLaneAI/pennylane/pull/775, https://github.com/PennyLaneAI/pennylane/pull/774, and https://github.com/PennyLaneAI/pennylane/pull/777 are merged, and a new version of PennyLane is released. The tutorial in this PR is currently statically rendered for feedback purposes, and can be merged when the outlined conditions are met.

**Title:**

Adds a tutorial that outlines all of the new functionality in the PennyLane QAOA module by walking the user through a basic example of QAOA for MinVertexCover.

**Summary:**

- Walks the user through the MinVertexcover problem
- Defines cost and mixer Hamiltonians
- Defines the QAOA layer
- Repeatedly applies the QAOA layer
- Optimizes the circuit and visualizes the results

**Relevant references:**

**Possible Drawbacks:**

**Related GitHub Issues:**

josh146 commented:
Hey @Lucaman99, you can make this PR against the `dev` branch, which is pinned to the GitHub master version of PennyLane :) Once the PRs are merged, it should build correctly

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+powerful features that allow for the QAOA module to be intuitive and
+modular.
+
+Defining Cost and Mixer Hamiltonians
```
I felt like there is some discontinuity in which the next couple sections introduce concepts. I think this is a good point at which to introduce the specific problem (vertex cover) - something for the reader to keep in mind going forward. However the idea of cost and mixer Hamiltonians (and Hamiltonians in general) isn't really introduced until later on, in the "constructing the QAOA layers" section. 

I'd recommend reshuffling the material into roughly this order:
 - Vertex cover (discussed in the abstract)
 - Discussion of unitary <-> Hamiltonian relationships
 - Make the transition to QAOA layers and the QAOA ansatz
 - Show how these are used to solve vertex cover - how to get the problem-specific Hamiltonians, optimization, etc.

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# Constructing the QAOA Layers
+# ----------------------------
+#
+# Oftentimes, when considering quantum circuits or variational ansatzae,
```
```suggestion
# Oftentimes, when considering quantum circuits or variational ansatze,
```

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# This unitary is determined completely in terms of a Hamiltonian,
+# :math:`\hat{H}` and a scalar :math:`t`. As it turns out, any unitary
+# :math:`U` can be written in the form :math:`e^{i H}`, where :math:`H` is
+# some Hamiltonian, making time-evolution the most fundamental and
```
I think more formally this is true for any Hermitian H (Hamiltonians are often Hermitian, but not always).

ixfoduap replied:
Yes, so we could change to "any unitary :math:`U` can be written in the form :math:`U=e^{i H}`, where :math:`H` is
some Hermitian operator, which we can interpret as a Hamiltonian."

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#           = \ \displaystyle\sum_{k} \hat{H}_k
+#
+# where :math:`U_{\text{Approx}}` approaches :math:`U` as :math:`n`
+# becomes larger. We can call this unitary with
```
I'd move this stuff about cost/mixer Hamiltonians under the next heading to tie it more closely to the ansatz discussion.

ixfoduap replied:
We can also give a short example of implementing a gate from a Hamiltonian with just one term using `ApproximateTimeEvolution`. We could then say that the reason that it is called this way is because for Hamiltonians with many terms, we use Trotter, and explain it right after

ixfoduap replied:
Then we give an example of implementing approx time evolution with some user-defined Hamiltonian with many terms

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# Due to this fact, we introduce the ``qml.layer`` function in PennyLane.
+# This general method allows us to take a function containing quantum
+# operations, a template, or even a single quantum gate and repeatedly
+# apply it to a set of wires.
```
Oh wow this is really cool functionality! :grin:

ixfoduap replied:
@glassnotes Do you like the name `qml.layer` or would you prefer `qml.repeat`?
This is an on-going existential question

glassnotes replied:
Actually I think I like `repeat` since it's  more descriptive and verb-like than `layer`. `tile1st could also work (like `np.tile`).

josh146 replied:
😆 I think I suggested `qml.tile` at one point as well

(@co9olguy @mariaschuld)

co9olguy replied:
Note, this question is not on-going. Time to move on :laughing:

ixfoduap replied:
:sob: :sob: :sob:

josh146 replied:
☠️ ☠️ ☠️

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+def circuit(params, **kwargs):
+
+    for w in wires:
+        qml.PauliX(wires=w)
```
To make the even superposition should this be Hadamard?

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+nx.draw(graph, pos, node_color=["r", "b", "r", "b"])
+plt.subplot(122)
+nx.draw(graph, pos, node_color=["b", "r", "r", "b"])
+plt.show()
```
Add concluding section to recap. Maybe also discuss additional/advanced features in the QAOA module that were beyond the scope of the tutorial?

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#
+#
+# In the case of QAOA, this allows us to easily define the ansatz. We
+# begin. with an even superposition over all basis states. Then, we layer
```
```suggestion
# begin with an even superposition over all basis states. Then, we layer
```

glassnotes commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# quantum neural networks, to Trotterization in quantum simulation.
+#
+#
+# .. figure:: ../demonstrations/qaoa_module/repeat.png
```
A caption or tooltip could be helpful here for indicating what these circuits are for.

ixfoduap replied:
This figure is great! :star_struck: But the circuits are appearing too small :disappointed: 
Maybe just take the first two (rotations + CNOTs and Grover) and stack them vertically? The last one is very similar ro the figure that follows

glassnotes reviewed (commented):
Cool tutorial @Lucaman99! I made some comments on the overall organization of the material, but in general I think the technical content provides a great starting point for people interested in using the module. 

(The graphs and stuff didn't render for me but I think that's related to the above point about other PRs not being merged?)

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# .. math:: H \ = \ H_1 \ + \ H_2 \ + \ H_3 \ + \ \cdots \ + \ H_N
+#
+# is not an easy task, and involves very long strings of quantum gates.
+# However, we are able to make use of the Trotter-Suzuki decomposition
```
```suggestion
# Instead, we are able to make use of the Trotter-Suzuki decomposition
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+    :align: center
+    :width: 90%
+
+The goal of this tutorial is to introduce PennyLane’s built-in QAOA
```
Actually, after going through the rest of the tutorial, I realize that this is a great spot to briefly describe the "philosophy" behind QAOA with PennyLane. We can say something like: "You will learn how to use time evolution and circuit repetitions to create QAOA circuits. Thus, PL provides general and flexible capabilities that can be tailored and refined to implement QAOA for a wide variety of problems"

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# However, we are able to make use of the Trotter-Suzuki decomposition
+# formula:
+#
+# .. math:: e^{A \ + \ B} \ \approx \ \Big(e^{A/n} e^{B/n}\Big)^{n} \ \ \ \ \ n \ \gg \ 1
```
```suggestion
# .. math:: e^{A \ + \ B} \ \approx \ \Big(e^{A/n} e^{B/n}\Big)^{n} 
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#
+# .. math:: e^{A \ + \ B} \ \approx \ \Big(e^{A/n} e^{B/n}\Big)^{n} \ \ \ \ \ n \ \gg \ 1
+#
+# To implement an *approximate* time-evolution unitary:
```
```suggestion
# to implement an *approximate* time-evolution unitary:
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#
+# To implement an *approximate* time-evolution unitary:
+#
+# .. math:: U_{\text{Approx}}(\hat{H}, t, n) \ = \ \displaystyle\prod_{j \ = \ 1}^{n}
```
```suggestion
# .. math:: U(\hat{H}, t, n) \ = \ \displaystyle\prod_{j \ = \ 1}^{n}
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#           \displaystyle\prod_{k} e^{-i \hat{H}_k t / n} \ \ \ \ \ \ \ \ \ \ \hat{H} \
+#           = \ \displaystyle\sum_{k} \hat{H}_k
+#
+# where :math:`U_{\text{Approx}}` approaches :math:`U` as :math:`n`
```
```suggestion
# where :math:`U` approaches :math:`e^{-i \hat{H} t}` as :math:`n`
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+
+######################################################################
+#
+# .. figure:: ../demonstrations/qaoa_module/layer.png
```
I'm not sure about the pink box in this figure...

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# .. math:: U_{\text{QAOA}}(\boldsymbol\gamma, \ \boldsymbol\alpha) \ = \ e^{-i \alpha_k H_M}
+#           e^{-i \gamma_k H_C} \ ... \ e^{-i \alpha_1 H_M} e^{-i \gamma_1 H_C}
+#
+# In other words, we are repetedly applying the ``qaoa_layer`` to a set of
```
```suggestion
# In other words, we are repeatedly applying the ``qaoa_layer`` to a set of
```
+1 cent :moneybag:

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#           e^{-i \gamma_k H_C} \ ... \ e^{-i \alpha_1 H_M} e^{-i \gamma_1 H_C}
+#
+# In other words, we are repetedly applying the ``qaoa_layer`` to a set of
+# wires. The idea of repetition is ubiquitous within quantum computing,
```
Nice!

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+
+    qml.layer(qaoa_layer, depth, params[0], params[1])
+
+
```
So this should be a new section, something like "Min vertex cover with QAOA"

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+#
+# Now that we have defined the full QAOA ansatz, we must define and
+# optimize the cost function. As is the case with most variational quantum
+# algorithms, the cost function we wish to minimize is simply the
+# expectation value of :math:`H_C`, with respect to the state prepared by
+# the ansatz. We can therefore use ``qml.VQECost`` to define this cost
```
```suggestion
# algorithms, the cost function is the
# expectation value of :math:`H_C` with respect to the state prepared by
# the circuit. We can therefore use ``qml.VQECost`` to define this cost
```

ixfoduap replied:
We also need to explain a bit more what `VQECost` is. From the name alone, which we should probably change, it's not clear what it does

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+
+
+# Defines the optimizer, steps, and initial parameters
+optimizer = qml.GradientDescentOptimizer()
```
Did we try other optimizers?

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# apply it to a set of wires.
+#
+#
+# .. figure:: ../demonstrations/qaoa_module/qml_layer.png
```
Love this figure!

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# covers of our graph:
+#
+
+
```
These plots are not showing for me

ixfoduap reviewed (commented):
Great work Jack! The foundation for a great tutorial is there. As suggested by Olivia, the most significant change is to re-organize the content and add some more flesh to the introduction and conclusion

Lucaman99 commented:
Thanks so much for the reviews @glassnotes @ixfoduap! I'll incorporate these changes ASAP!

josh146 commented:
@Lucaman99: you can now convert this tutorial into an executable one (it should be as simple as changing the filename to begin with `tutorial_`, and updating the TOC in `demonstrations.rst`).

It should work fine, since the `PennyLaneAI:dev` branch uses the GitHub master of PennyLane to build :) (@glassnotes @ixfoduap this will cause the plots to render)

co9olguy commented:
Awesome, when might this be ready to publish and share?

Lucaman99 commented:
Hey everyone! Sorry for the delay, the past couple weeks have been incredibly busy at school. I'm in the process of addressing these comments/TODOs, as well as writing the code for the Hamiltonian arithmetic section. Hopefully this will all be done this week 🙂

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+these circuits to create more powerful algorithms. These simple ingredients,
+together with the ability to optimize quantum circuits, are the building blocks of QAOA. By focusing
+on the fundamentals, PennyLane provides general and flexible capabilities that can be tailored and
+refined to implement QAOA for a wide variety of problems. In the last of the tutorial, you will
```
```suggestion
refined to implement QAOA for a wide variety of problems. In the last part of the tutorial, you will
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,298 @@
+# it is useful to think of a quantum circuit in terms of a *Hamiltonian*.
+#
+#
+# .. figure:: ../demonstrations/qaoa_module/ham_circuit.png
```
New figure is great, but the centering is off...

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+which is a unitary defined as:"""
+
+######################################################################
+#  .. math:: U(\hat{H}, \ t) \ = \ e^{-i \hat{H} t / \hbar}.
```
Let's drop the hats from the Hamiltonians. We don't use it in other parts and it makes things look more clunky

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# In general, implementing a quantum circuit that exactly exponentiates a Hamiltonian
+# with many non-commuting terms, i.e., a Hamiltonian of the form:
+#
+# .. math:: H \ = \ H_1 \ + \ H_2 \ + \ H_3 \ + \ \cdots \ + \ H_N
```
```suggestion
# .. math:: H \ = \ H_1 \ + \ H_2 \ + \ H_3 \ + \ \cdots \ + \ H_N,
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# becomes larger.
+#
+# In PennyLane, this is implemented using the ``qml.templates.ApproxTimeEvolution`` template.
+# For example,  (TODO: add  example, with circuit printout)
```
This is still left to do

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# from amplitude amplification in `Grover’s algorithm
+# <https://en.wikipedia.org/wiki/Grover%27s_algorithm>`__
+# to layers in `quantum neural networks
+# <https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.1.033063>`__,
```
```suggestion
# <https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.1.033063>`__
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+#
+# To create a larger circuit consisting of many repetitions, we pass the circuit to be
+# repeated as an argument and specify the number of repetitions:
+# (TODO: add code snippet with circuit printout)
```
This is still left to do...

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+#     :width: 90%
+#
+# We have learnt how time evolution can be used to create circuits from Hamiltonians,
+# and how these can be layered to create longer circuits. This means we are now ready to
+# explore QAOA.
```
```suggestion
# and how these can be layered to create longer circuits. We are now ready to
# explore QAOA.
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+#
+# 4. Choose a parameter :math:`n\geq 1` and build the circuit
+#
+#    .. math:: U_{\text{QAOA}}(\boldsymbol\gamma, \ \boldsymbol\alpha) \ = \ e^{-i \alpha_n H_M}
```
```suggestion
#    .. math:: U(\boldsymbol\gamma, \ \boldsymbol\alpha) \ = \ e^{-i \alpha_n H_M}
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# cover <https://en.wikipedia.org/wiki/Vertex_cover>`__ of a graph.
+# A vertex cover is a collection of vertices such that
+# each edge in the graph contains at least one of the vertices in the cover. Hence,
+# these vertices "cover" all the edges.
```
```suggestion
# these vertices "cover" all the edges 👍.
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# We wish to find the vertex cover that has the
+# smallest possible number of vertices. Vertex covers can be represented by a bit string
+# where each bit denotes whether the corresponding vertex is present in the cover. For example,
+# the bit string 1001 represents a cover consisting of the first and last vertex.
```
```suggestion
# the bit string 01010 represents a cover consisting of the second and fourth vertex in a graph with five vertices.
```
Adding this to match the figure below

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+import networkx as nx
+
+
+######################################################################
```
Spacing seems a bit off in the built doc for this part

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# with the solutions of the problem. The mixer Hamiltonian is the simple, non-commuting sum of Pauli-X
+# operations on each node of the graph:
+
+cost_h, mixer_h = qaoa.min_vertex_cover(graph, contrained=False)
```
```suggestion
cost_h, mixer_h = qaoa.min_vertex_cover(graph, constrained=False)
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+
+######################################################################
+# A single layer of QAOA consists of time evolution under these
+# Hamiltonians. While it is possible to use `ApproxTimeEvolution`, the QAOA module allows you to
```
```suggestion
# Hamiltonians. While it is possible to use ``ApproxTimeEvolution``, the QAOA module allows you to
```
We also need to add links to the documentation for these functions

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+depth = 2
+
+
+# Defines the full QAOA circuit
```
Remove all comments from this and subsequent code snippets. This section is looking a bit too heavy now

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+
+######################################################################
+# The states
+# :math:`|10\rangle \ = \ |1010\rangle` and
+# :math:`|6\rangle \ = \ |0110\rangle` have the highest probabilities of
```
```suggestion
# :math:`|6\rangle \ = \ |0110\rangle` and
# :math:`|10\rangle \ = \ |1010\rangle` have the highest probabilities of
```
10 = 1010 is one of my favourite equations

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# favour :math:`|6\rangle`, making it the only ground state.
+#
+# It is easy to introduce constraints of this form in PennyLane. We can use the ``qml.qaoa.edge_driver`` cost
+# Hamiltonian to "reward" cases in which a first and last vertices of the graph
```
```suggestion
# Hamiltonian to "reward" cases in which the first and last vertices of the graph
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+# Hamiltonian to "reward" cases in which a first and last vertices of the graph
+# are :math:`0`:
+
+cost_term = qaoa.edge_driver(nx.Graph([(0, 3)]), ['00'])
```
`cost_term` is a bit strange. How about `reward_h`?

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+cost_term = qaoa.edge_driver(nx.Graph([(0, 3)]), ['00'])
+
+######################################################################
+# Then, we can weight this now constraining term appropriately, and add
+# it to the original minimum vertex cover Hamiltonian:
```
This reads awkwardly, consider revising

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+new_cost_h = cost_h + 2*cost_term
+
+######################################################################
+# Notice that PennyLane allows for simple addition and multiplication
```
```suggestion
# Notice that PennyLane allows for simple addition and multiplication of
```

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -0,0 +1,416 @@
+#
+
+######################################################################
+# TODO: Add conclusion
```
Still left to do

ixfoduap reviewed (commented):
The tutorial is looking so good!

Please address the few comments I'm leaving here and all parts that are left TODO

josh146 commented:
@Lucaman99, just two small comments:

* Don't forget to make the tutorial executable! I just had a check, and there is still no output.

* We recently re-arranged our demonstration pages, so there will be quite a significant merge conflict :) You will need to overwrite `demonstrations.rst` from `master`, and move the thumbnail for this demo to `demos_basics.rst` in the 'Applications' section.

ixfoduap commented on demonstrations/qaoa_mod.py:
```diff
@@ -452,5 +438,15 @@ def probability_circuit(gamma, alpha):
+# Conclusion
+# ----------
+#
+# In summary, this tutorial has demonstrated how to use the PennyLane QAOA functionality, while
+# also surveying some of the fundmental PennyLane features that make the QAOA module so simple and
+# flexible. Now, it's your turn to experiment with QAOA! If you need some inspiration for how to get
+# started:
```
```suggestion
# You have learnt how to use the PennyLane QAOA functionality, while
# also surveying some of the fundamental features that make the QAOA module simple and
# flexible. Now, it's your turn to experiment with QAOA! If you need some inspiration for how to get
# started:
```

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+#    consisting of repeated application of the cost and mixer layers.
+#
+# 5. Prepare an initial state, apply :math:`U(\boldsymbol\gamma,\boldsymbol\alpha)`,
+#    and use classical techniques, e.g. gradient descent, to optimize the parameters.
```
```suggestion
#    and use classical techniques to optimize the parameters.
```

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+#
+# In summary, the starting point of QAOA is the specification of cost and mixer Hamiltonians.
+# We then use time evolution and layering to create a variational circuit and optimize its
+# parameters. Finally, we sample from the circuit to get an approximate solution to
+# the optimization problem. Let's see it in action! 🚀
```
```suggestion
# parameters. The algorithm concludes by sampling from the circuit to get an approximate solution to
# the optimization problem. Let's see it in action! 🚀
```

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+# ------------------------------
+#
+# Our goal is to find the `minimum vertex
+# cover <https://en.wikipedia.org/wiki/Vertex_cover>`__ of a graph.
+# A vertex cover is a collection of vertices such that
```
```suggestion
# cover <https://en.wikipedia.org/wiki/Vertex_cover>`__ of a graph:
# a collection of vertices such that
```

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+# each edge in the graph contains at least one of the vertices in the cover. Hence,
+# these vertices "cover" all the edges 👍.
+# We wish to find the vertex cover that has the
+# smallest possible number of vertices. Vertex covers can be represented by a bit string
```
Insert new paragraph starting with "Vertex covers can..."

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+
+
+######################################################################
+# We also define the four-vertex graph for which we
```
This line of text also appears very squeezed in the docs, consider adding vertical space

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+# We also define the four-vertex graph for which we
+# want to find the minimum vertex cover:
+
+# Defines the graph
```
Here and below, remove code comments

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+
+cost_h, mixer_h = qaoa.min_vertex_cover(graph, constrained=False)
+
+print("Cost Hamiltonian")
```
Will need to see how this looks printed out, but consider removing the `print("---------------")` line

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+    qaoa.mixer_layer(alpha, mixer_h)
+
+######################################################################
+# We are now ready to build the full variational circuit. The number of wires is equal to
```
From here onwards, the text and code looks crammed. I suggest adding vertical space

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+#
+# QAOA is not one-size-fits-all when it comes to solving optimization problems. In many cases,
+# cost and mixer Hamiltonians will be very specific to one scenario, and not necessarily
+# fit within the mold of the pre-defined problems in the ``qml.qaoa`` submodule. Luckily,
```
```suggestion
# fit within the structure of the pre-defined problems in the ``qml.qaoa`` submodule. Luckily,
```
mold makes me think of the stuff that grows on rotten food

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+# QAOA submodule as well!
+#
+# The QAOA workflow above gave us two optimal solutions: :math:`|6\rangle = |0110\rangle`
+# and :math:`|10\rangle = |1010\rangle`. However, what if we were to add an addition constraint
+# that made one of these solutions "better" than the other? Let's imagine that we are interested in
```
```suggestion
# and :math:`|10\rangle = |1010\rangle`. What if we add a constraint
# that made one of these solutions "better" than the other? Let's imagine that we are interested in
```

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,452 @@
+
+######################################################################
+# Notice that PennyLane allows for simple addition and multiplication of
+# Hamiltonian objects using inline arithmetic operations! Finally, we can
```
```suggestion
# Hamiltonian objects using inline arithmetic operations! ➕ ➖ ✖️➗ Finally, we can
```

ixfoduap reviewed (commented):
Main comment: the last part of the tutorial looks a bit crammed. I think it will look better with more vertical space between text and code

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+
+plt.style.use("seaborn")
+plt.bar(range(2 ** len(wires)), probs)
+plt.show()
```
This is what I'm seeing in the built doc, which suggests that the optimization is not working properly
![Screenshot from 2020-10-14 09-24-35](https://user-images.githubusercontent.com/40441298/95995904-0f097900-0e00-11eb-8c05-33ba8eec8721.png)

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+# being measured, just as expected!
+
+
+pos = nx.spring_layout(graph)
```
Instead of creating this plot inline, let's save it and upload it directly

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+#
+# QAOA is not one-size-fits-all when it comes to solving optimization problems. In many cases,
+# cost and mixer Hamiltonians will be very specific to one scenario, and not necessarily
+# fit within the structure of the pre-defined problems in the ``qml.qaoa`` submodule. Luckily,
```
Add actual link to the qaoa submodule

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+
+for i in range(steps):
+    params = optimizer.step(cost_function, params)
+    print("Step {} / {}".format(i + 1, steps))
```
As above, probably best to remove this print statement

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+# - Experiment with different optimizers and different devices. Which ones work the best?
+# - Play around with some of the other built-in cost and mixer Hamiltonians.
+# - Try making your own custom constraining terms. Is QAOA properly amplifying some bitstrings over others?
+#
```
I think it would look great to end with the same image that we started with, the QAOA circuit with red and blue boxes

trbromley commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+#     :align: center
+#     :width: 70%
+#
+# In PennyLane, this is implemented using the `~.pennylane.templates.ApproxTimeEvolution` template.
```
```suggestion
# In PennyLane, this is implemented using the :func:`~.pennylane.templates.ApproxTimeEvolution` template.
```

trbromley commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+#     :width: 100%
+#
+#
+# Circuit repetition is implemented in PennyLane using the `~.pennylane.layer` function. This method
```
```suggestion
# Circuit repetition is implemented in PennyLane using the :func:`~.pennylane.layer` function. This method
```

trbromley reviewed (commented):
Thanks @Lucaman99 it looks awesome!

Do we need the `.DS_Store` file?

Lucaman99 commented:
🤔 Hmmmm, I'm not sure why the probability distributions look so weird. I ran it a few times locally and it seems to work fine. I'll keep digging!

Lucaman99 commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+# We then weigh the constraining term appropriately and add
+# it to the original minimum vertex cover Hamiltonian:
+
+new_cost_h = cost_h + 2*reward_h
```
I changed this to 0.75, essentially because we need to choose a weight that is not too high, or else bitstrings that satisfy only the constraint will occur with a higher probability, but not too low that it has no effect. Empirically, this seems to work the best.

Lucaman99 commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,470 @@
+######################################################################
+#
+# Just as we expected, the :math:`|6\rangle` state is now favoured
+# over :math:`|10\rangle`!
```
This should hopefully be resolved now!

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,472 @@
+
+######################################################################
+#
+# We have learnt how time evolution can be used to create circuits from Hamiltonians,
```
```suggestion
# We have learned how time evolution can be used to create circuits from Hamiltonians,
```

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,472 @@
+# 1. Define a *cost Hamiltonian* :math:`H_C` such that its ground state
+#    encodes the solution to the optimization problem.
+#
+# 2. Define a *mixer Hamiltonian* :math:`H_C`.
```
```suggestion
# 2. Define a *mixer Hamiltonian* :math:`H_M`.
```

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,472 @@
+.. meta::
+    :property="og:description": Learn how to implement QAOA with PennyLane
+    :property="og:image": https://pennylane.ai/qml/_images/qaoa_layer.png
+
```
```suggestion
.. related::

   tutorial_qaoa_maxcut QAOA for MaxCut
```

glassnotes commented on demos_basics.rst:
```diff
@@ -131,6 +131,12 @@ hardware.
     :description: :doc:`demos/tutorial_vqe_uccsd_obs`
     :tags: chemistry
+    
+.. customgalleryitem::
```
Should this maybe be moved to right before the other QAOA tutorial so that the introductory one comes before the application to MaxCut?

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,489 @@
+for i in range(steps):
+    params = optimizer.step(cost_function, params)
+
+print("Optimal Parameters: {}".format(params))
```
The print statement here leads to an output with a horizontal scroll bar - to avoid this, perhaps have the two sets of parameters be printed on separate lines, or alternatively print with reduced precision (like 6 decimal points or something).

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,489 @@
+# :math:`|10\rangle \ = \ |1010\rangle` have the highest probabilities of
+# being measured, just as expected!
+#
+# .. figure:: ../demonstrations/qaoa_module/graph.png
```
Label vertices to match earlier graphic.

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,489 @@
+# *but also colour the first and third vertices* :math:`1`. A constraint of this form will
+# favour :math:`|10\rangle`, making it the only true ground state.
+#
+# It is easy to introduce constraints of this form in PennyLane. We can use the :func:`~.qaoa.edge_driver` cost
```
Provide link to the `edge_driver` docs - it took me a sec to understand the signature so would be helpful to the info handy.

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,489 @@
+for i in range(steps):
+    params = optimizer.step(cost_function, params)
+
+print("Optimal Parameters: {}".format(params))
```
Same as above, print on two lines or reduce precision.

glassnotes commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,489 @@
+
+######################################################################
+#
+# Just as we expected, the :math:`|10\rangle` state is now favoured
```
Nice!! :tada: :star:

glassnotes reviewed (approved):
Looks awesome @Lucaman99! Just a few minor comments, but otherwise the results are great and I think this is ready to be merged in.

glassnotes commented:
> Looks awesome @Lucaman99! Just a few minor comments, but otherwise the results are great and I think this is ready to be merged in.

I could not figure out how to comment about it in the review, but there is a `.DS_Store` in there that should be removed.

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -318,19 +326,33 @@ def circuit(params, **kwargs):
+
+######################################################################
+#
+# Notice that we set each of the initial parameters to :math:`0.5`. For demonstration purposes,
+# we chose initial parameters that we know work fairly well, and don't get stuck in any local minima.
+#
+# The choice of initial parameters for a variational circuit is usually a difficult problem,
+# so we won't linger on it too much in this tutorial, but it is important to note that
+# finding an initial set of parameters that work well for a few toy problems often yields good results
+# for more complex instances of the algorithm as well.
```
```suggestion
# Notice that we set each of the initial parameters to :math:`0.5`. For demonstration purposes,
# we chose initial parameters that work fairly well and don't get stuck in any local minima.
# The choice of initial parameters for a variational circuit is usually a difficult problem,
# potentially requiring repeated initialization of a circuit until good performance is observed. 

```

ixfoduap commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -0,0 +1,491 @@
+# we chose initial parameters that work fairly well and don't get stuck in any local minima.
+# The choice of initial parameters for a variational circuit is usually a difficult problem,
+# potentially requiring repeated initialization of a circuit until good performance is observed. 
+
```
```suggestion
```
